unittests/BuildSystem/BuildExecutionQueueTest.cpp

```diff
@@ -0,0 +1,112 @@
+
+//TODO: Remove unused headers.
+#include "llbuild/Core/BuildDB.h"
+#include "llbuild/Core/BuildEngine.h"
+#include "llbuild/Core/DependencyInfoParser.h"
+#include "llbuild/Core/MakefileDepsParser.h"
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llbuild/BuildSystem/BuildExecutionQueue.h"
+#include "llbuild/BuildSystem/ExternalCommand.h"
+#include "llvm/Support/raw_ostream.h"
+#include "gtest/gtest.h"
+
+#include "llbuild/BuildSystem/BuildSystem.h"
+#include "llbuild/BuildSystem/BuildSystemCommandInterface.h"
+
+#include <unordered_map>
+#include <vector>
+
+using namespace llbuild;
+using namespace llbuild::core;
+using namespace llbuild::buildsystem;
+
+namespace {
+class PhonyCommand : public ExternalCommand {
+  public:
+    using ExternalCommand::ExternalCommand;
+
+    virtual bool shouldShowStatus() override { return false; }
+
+    virtual void getShortDescription(SmallVectorImpl<char> &result) override {
+      llvm::raw_svector_ostream(result) << getName();
+    }
+
+    virtual void getVerboseDescription(SmallVectorImpl<char> &result) override {
+      llvm::raw_svector_ostream(result) << getName();
+    }
+
+    virtual bool executeExternalCommand(BuildSystemCommandInterface& bsci,
+        Task* task,
+        QueueJobContext* context) override {
+      // Nothing needs to be done for phony commands.
+      llvm::outs() << "HEY BRO";
+
+      return true;
+    }
+};
+  
+class SimpleBuildExecutionQueueDelegate : public BuildExecutionQueueDelegate {
+
+  public:
+    SimpleBuildExecutionQueueDelegate() { 
+    }
+
+    virtual void commandJobStarted(Command* command) override {
+    }
+
+    virtual void commandJobFinished(Command* command) override {
+    }
+
+    virtual void commandProcessStarted(Command* command,
+        ProcessHandle handle) override {
+    }
+
+    virtual void commandProcessHadError(Command* command, ProcessHandle handle,
+        const Twine& message) override {
+    }
+
+    virtual void commandProcessHadOutput(Command* command, ProcessHandle handle,
+        StringRef data) override {
+    }
+
+    virtual void commandProcessFinished(Command* command, ProcessHandle handle,
+        int exitStatus) override {
+    }
+};
+
+TEST(BuildSystemTest, basic) {
+  SimpleBuildExecutionQueueDelegate delegate;
+  std::shared_ptr<BuildExecutionQueue> queue = std::shared_ptr<BuildExecutionQueue>(
+      createLaneBasedExecutionQueue(delegate, 1));
+  auto phony = llvm::make_unique<PhonyCommand>("Phony");
+  auto fn = [&](QueueJobContext* context) {
+    queue->executeProcess(context, {"/Users/aciid/mycode/exe/a.out"}, {}, 
```

Comment:
The a.out here won't work obviously. :)

My recommendation is that if we need a custom tool for probing command output behaviors, we write a little Python or C++ utility helper which takes command line arguments telling it to fake various behaviors (e.g., randomly crashing, or writing a bunch of output, or closing one pipe and not the other, or spawning a subprocess which keeps the pipe open while the command itself completes early).